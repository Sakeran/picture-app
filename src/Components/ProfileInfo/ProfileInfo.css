.ProfileInfo{
  background-color: white;
  margin-bottom: 1em;
  text-align: center;
}

.ProfileInfo-name {
  margin-top: 1em;
}

.ProfileInfo-section {
  margin: 1em 0;
}
