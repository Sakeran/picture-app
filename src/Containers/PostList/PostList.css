.PostList-masonry {
  margin: 0 auto;
}
