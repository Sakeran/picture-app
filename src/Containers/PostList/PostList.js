import React from 'react';
import { connect } from 'react-redux';

import PostCard from '../../Components/PostCard/PostCard';
import getPost from '../../GraphQL/getPost';

class PostList extends React.Component {

  constructor(props) {
    super(props);
    this.state = { loading: true };
  }

  componentDidMount() {
    // Check each post id as it is passed in. Query for any data that does
    // not exist in store. The list will only display once everything is
    // finished loading.
    const queries = this.props.postIDs.map(e => {
      if(!this.props.posts[e]) {
        return getPost(e)
        .then(post => {
          this.props.addPost(post);
        });
      }
      return Promise.resolve(true);
    });
    Promise.all(queries)
    .then(() => {
      this.setState({ loading: false });
    });
  }

  render() {
    return (
      <div>
      <h3> {this.state.loading ? "Loading" : "Ready!"} </h3>
        {this.props.postIDs.map(e => (
          <PostCard key={e} post={this.props.posts[e]} />
        ))}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  posts: state.posts
});

const mapDispatchToProps = (dispatch) => ({
  addPost: (post) => dispatch({type:'ADD_POST', post: post})
});

export { PostList }
export default connect(mapStateToProps, mapDispatchToProps)(PostList);
