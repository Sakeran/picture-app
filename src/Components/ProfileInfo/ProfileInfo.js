import React from 'react';
import PropTypes from 'prop-types';

import './ProfileInfo.css';

const ProfileInfo = ({ name, location, bio}) => (
  <div className="ProfileInfo section-border">
    <h2 className="ProfileInfo-name">{name || "No Name Given"}</h2>
    <div className="ProfileInfo-section">
      <strong>Location:</strong> {location}
    </div>
    <div className="ProfileInfo-section">
      <strong>About:</strong> {bio}
    </div>
  </div>
);

ProfileInfo.propTypes = {
  name: PropTypes.string,
  location: PropTypes.string,
  bio: PropTypes.string,
};

export default ProfileInfo;
