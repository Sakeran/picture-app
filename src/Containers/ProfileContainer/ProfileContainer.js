import React from 'react';
import PropTypes from 'prop-types';

import { graphql } from 'react-apollo';
import userProfileAndPosts from '../../GraphQL/Queries/userProfileAndPosts';

import ProfileInfo from '../../Components/ProfileInfo/ProfileInfo';
import PostList from '../PostList/PostList';

class ProfileContainer extends React.Component {
  render() {
    const {data: {loading, error, user }, loadMorePosts } = this.props;
    if (loading) {
      return <h2 className="header">Loading...</h2>;
    }
    if (error) {
      return <h2 className="header">Error while loading profile.</h2>;
    }
    const postListProps = {
      posts: user.posts,
      totalPosts: user.postCount,
      loadMorePosts
    };
    return (
      <div>
        <ProfileInfo {...user.profile} />
        {
          user.postCount ?
            (<div>
              <h2 className="header">This user has the following posts:</h2>
              <PostList {...postListProps} />
            </div>)
          :
            <h2 className="header">This user has not posted anything yet.</h2>
        }
      </div>
    )
  }
}

ProfileContainer.propTypes = {
  userId: PropTypes.string.isRequired,
};

export { ProfileContainer };
export default graphql(userProfileAndPosts, {
  options: (props) => ({
    variables: {
      id: props.userId,
      postOffset: 0,
    }
  }),
  props: (props) => ({
    ...props,
    loadMorePosts: () => {
      return props.data.fetchMore({
        variables: {
          id: props.ownProps.userId,
          postOffset: props.data.user.posts.length
        },
        updateQuery: (prevResult, { fetchMoreResult }) => {
          if (!fetchMoreResult) { return prevResult; }
          return Object.assign({}, prevResult, {
            user: Object.assign({}, prevResult.user, {
              posts: [...prevResult.user.posts, ...fetchMoreResult.user.posts]
            })
          });
        }
      })
    }
  })
})(ProfileContainer);
