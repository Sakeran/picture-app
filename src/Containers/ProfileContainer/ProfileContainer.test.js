import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import {ProfileContainer} from './ProfileContainer';
import {PostList} from '../PostList/PostList';
import ProfileInfo from '../../Components/ProfileInfo/ProfileInfo';

const makePost = (id) => ({
  id,
  image: 'http://example.com/img.png',
  likeCount: 0,
  commentCount: 0,
});

const testPosts = [1,2,3].map(i => makePost(i));

test('Renders properly while loading', () => {
  const containerProps = {
    userId: '11111',
    data: {
      loading: true
    }
  };
  const component = shallow(<ProfileContainer {...containerProps} />);
  expect(component.find(ProfileInfo)).toHaveLength(0);
  expect(component.find(PostList)).toHaveLength(0);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

test('Renders properly with an error', () => {
  const containerProps = {
    userId: '11111',
    data: {
      loading: false,
      error: true,
    }
  };
  const component = shallow(<ProfileContainer {...containerProps} />);
  expect(component.find(ProfileInfo)).toHaveLength(0);
  expect(component.find(PostList)).toHaveLength(0);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

test('Renders properly when loaded', () => {
  const containerProps = {
    userId: '11111',
    data: {
      loading: false,
      error: false,
      user: {
        profile: {
          name: 'Test User',
          location: 'Somewhere',
          bio: 'A test user',
        },
        posts: testPosts,
        postCount: testPosts.length
      }
    }
  };
  const component = shallow(<ProfileContainer {...containerProps} />);
  expect(component.find(ProfileInfo)).toHaveLength(1);
  expect(component.find(PostList)).toHaveLength(1);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
