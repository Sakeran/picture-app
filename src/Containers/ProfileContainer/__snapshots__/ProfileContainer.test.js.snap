// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders properly when loaded 1`] = `
<div>
  <ProfileInfo
    bio="A test user"
    location="Somewhere"
    name="Test User"
  />
  <div>
    <h2
      className="header"
    >
      This user has the following posts:
    </h2>
    <PostList
      posts={
        Array [
          Object {
            "commentCount": 0,
            "id": 1,
            "image": "http://example.com/img.png",
            "likeCount": 0,
          },
          Object {
            "commentCount": 0,
            "id": 2,
            "image": "http://example.com/img.png",
            "likeCount": 0,
          },
          Object {
            "commentCount": 0,
            "id": 3,
            "image": "http://example.com/img.png",
            "likeCount": 0,
          },
        ]
      }
      totalPosts={3}
    />
  </div>
</div>
`;

exports[`Renders properly while loading 1`] = `
<h2>
  Loading...
</h2>
`;

exports[`Renders properly with an error 1`] = `
<h2>
  Error while loading profile.
</h2>
`;
